### PR TITLE
builtin: IsIPv4Mapped function

### PR DESCRIPTION
**Pull request by qhsong**
Add builtin function `IsIPv4Mapped`

### CODE REVIEW

File: expression/builtin_miscellaneous.go [hunk 1]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+	if arg.IsNull() {
+		return d, nil
+	}
+
+	ipAddress, err := arg.ToBytes()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	if len(ipAddress) != net.IPv6len {
```
Comment (shenli) [inline expression/builtin_miscellaneous.go]: mysql> select IS_IPV4_MAPPED("127.0.0.1");
+-----------------------------+
| IS_IPV4_MAPPED("127.0.0.1") |
+-----------------------------+
|                           0 |
+-----------------------------+
1 row in set (0.04 sec)

mysql> 

MySQL returns 0 for IPv4 address, not null.

File: expression/builtin_miscellaneous.go [hunk 2]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+		return d, nil
+	}
+
+	for _, bytes := range ipAddress[:10] {
+		if bytes != 0x0 {
+			return d, nil
+		}
+	}
+
+	if ipAddress[10] != 0xff && ipAddress[11] != 0xff {
```
Comment (shenli) [inline expression/builtin_miscellaneous.go]: if ipAddress[10] != 0xff || ipAddress[11] != 0xff ？

**Comment by shenli** [comment]
@XuHuaiyu @tiancaiamao PTAL

File: expression/builtin_miscellaneous.go [hunk 3]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
 // eval evals a builtinIsIPv4MappedSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_is-ipv4-mapped
 func (b *builtinIsIPv4MappedSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("IS_IPV4_MAPPED")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return types.Datum{}, errors.Trace(err)
```
Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: return d, errors.Trace(err)

File: expression/builtin_miscellaneous.go [hunk 4]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+	}
+
+	arg := args[0]
+	d.SetInt64(0)
+	if arg.IsNull() {
+		return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: s/return d, nil/return

File: expression/builtin_miscellaneous.go [hunk 5]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+	}
+
+	ipAddress, err := arg.ToBytes()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	if len(ipAddress) != net.IPv6len {
+		//Not an IPv6 address, return false
+		return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: ditto

File: expression/builtin_miscellaneous.go [hunk 6]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+	}
+
+	for _, bytes := range ipAddress[:10] {
+		if bytes != 0x0 {
+			return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: ditto

File: expression/builtin_miscellaneous.go [hunk 7]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+		}
+	}
+
+	if ipAddress[10] != 0xff && ipAddress[11] != 0xff {
+		return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: ditto

File: expression/builtin_miscellaneous.go [hunk 8]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+	}
+
+	d.SetInt64(1)
+	return d, nil
```
Comment (XuHuaiyu) [inline expression/builtin_miscellaneous.go]: ditto

File: expression/builtin_miscellaneous.go (hunk 2 above)
Comment (qhsong) [inline expression/builtin_miscellaneous.go]: fix it

**Comment by qhsong** [comment]
@shenli @XuHuaiyu  fixed as you comment, pls review it.

**Comment by XuHuaiyu** [comment]
LGTM
PTAL @shenli @coocood

File: expression/builtin_miscellaneous.go [hunk 9]
```diff
@@ -525,7 +525,39 @@ type builtinIsIPv4MappedSig struct {
...
+	d.SetInt64(0)
+	if arg.IsNull() {
+		return d, nil
+	}
+
+	ipAddress, err := arg.ToBytes()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
```
Comment (tiancaiamao) [inline expression/builtin_miscellaneous.go]: I think you can try this, much simpler:
```
var v4InV6Prefix = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff}
if !bytes.HasPrefix(ipAddress, v4InV6Prefix) {
    return
}
d.SetInt64(1)
return
```

File: expression/builtin_miscellaneous_test.go [hunk 10]
```diff
@@ -221,3 +221,31 @@ func (s *testEvaluatorSuite) TestInet6AtoN(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(r.IsNull(), IsTrue)
 }
+
+func (s *testEvaluatorSuite) TestIsIPv4Mapped(c *C) {
+	tests := []struct {
+		ip     []byte
+		expect interface{}
+	}{
```
Comment (tiancaiamao) [inline expression/builtin_miscellaneous_test.go]: The test cases is not intuitive, if would be better if you add something like:

```
{[]byte(net.ParseIP("::10.0.5.9")), 0}
{[]byte(net.ParseIP("::ffff:192.168.0.1"),  1}
{[]byte(net.ParseIP('"::ffff:c0a8:0001"), 1}
{[]byte(net.ParseIP("::ffff:c0a8:1"), 1}
```

File: expression/builtin_miscellaneous.go (hunk 9 above)
Comment (qhsong) [inline expression/builtin_miscellaneous.go]: Much simple than me, do it later

File: expression/builtin_miscellaneous_test.go (hunk 10 above)
Comment (qhsong) [inline expression/builtin_miscellaneous_test.go]: Nope, we can not do that, cause `net.ParseIP`  parse  a mapped IPv6 address into its origin IPv4 address. It not works as expected. So we need directly use `[]byte` to do that.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**